Remove dependency on Python

augi opened the pull request:
A new JVM with small program is executed instead of Python.

The test is passing on Windows.

thesamet commented on src/main/scala/protocbridge/frontend/WindowsPluginFrontend.scala:
```diff
@@ -31,30 +33,28 @@ class WindowsPluginFrontend(pythonExecutable: String) extends PluginFrontend {
     val batchFile = PluginFrontend.createTempFile(".bat",
       s"""@echo off
-          |$pythonExecutable -u $pythonScript $port
+          |"${sys.props("java.home")}\\bin\\java.exe" -cp "${sys.props("java.class.path")}" ${WindowsPluginFrontend.getClass.getName.stripSuffix("$")} $port
```
What is the value of `java.class.path`? I am concerned that this wouldn't work when class loaders are involved - I am not sure if `java.class.path` would be affected. Were you able to test it when it's invoked as an sbt plugin to see if it works?

We could have a safer alternative that doesn't require `java.class.path`: how about rewriting 
`WindowsPluginFrontend` as a pure Java program so it doesn't have any dependencies (not even on Scala)? Then `java.exe` invocation would be simpler.

augi replied:
`java.class.path` contains the class path that was used to start the JVM - so it will not obviously work in case of `sbt` because the plugins are loaded at runtime 😞
Maybe `getClassLoader.getURLs()` would be sufficient because it returns also urls appended at runtime.

I will look at how the sbt plugins work and investigate that. It will take some time because I have no experience with sbt.

thesamet replied:
protoc-bridge is independent on SBT, though it sometimes runs inside. If you write it in pure Java, then you can invoke `java.exe` and provide it with the full path to the class file: https://stackoverflow.com/a/4546650/97524 (or maybe just the directory path with `-cp`)

augi replied:
Yep, I will rewrite it to pure Java anyway. The hardest work will be to test it in sbt context properly IMHO.

thesamet replied:
Thanks. For testing your work in SBT: `sbt-protoc` has tests that will exercise the code path that you are working on. You can use the following flow to test your work:
1. In this project run `+publishLocal` in SBT. 
2. Clone `sbt-protoc` and change the dependency on protoc-bridge to the SNAPSHOT version you published locally.
3. Make the necessary edits to `sbt-protoc` to make it compile.
4. in `sbt-protoc` SBT session, run `^scripted`.

augi replied:
Ok, I wasn't able to make `sbt-protoc` working on Windows but I did this:
1. I added a code to `WindowsPluginFrontend` class body (~ctor) to print what would be printed to the `bat` file.
1. `publishM2` in `protoc-bridge`.
1. Execute `^scripted` in scope of `sbt-protoc` project.
1. I saw the printed command, in my case `"C:\Program Files\Java\jdk1.8.0_151\jre\bin\java.exe" -cp "/C:/Users/augustyn/.m2/repository/com/thesamet/scalapb/protoc-bridge_2.10/0.7.1-SNAPSHOT/protoc-bridge_2.10-0.7.1-SNAPSHOT.jar" protocbridge.frontend.BridgeApp`
1. I tried to execute this command and it ended with this output:
```
Exception in thread "main" java.lang.ArrayIndexOutOfBoundsException: 0
        at protocbridge.frontend.BridgeApp.main(BridgeApp.java:14)
```

This is expected and demonstrates that the application was successfully executed.

So I believe it's in working state now.

thesamet replied:
Can you elaborate on `sbt-protoc` not working on windows? What did you try and what did you see? We need the `^scripted` tests to pass see we know it's working end-to-end.  `sbt-protoc` also has an `examples` directory with a few projects you can try.

thesamet commented on src/main/scala/protocbridge/frontend/WindowsPluginFrontend.scala:
```diff
@@ -1,22 +1,24 @@
+  * is a batch script that invokes WindowsPluginFrontend.main method, in a new JVM with the same parameters
+  * as the currently running JVM. The plugin will communicate its stdin and stdout to this socket.
   */
-class WindowsPluginFrontend(pythonExecutable: String) extends PluginFrontend {
```
For backward binary compatibility, can you add a deprecated constructor that accepts a string and drops it?

augi replied:
Done.

augi commented:
@thesamet I tried to execute `^scripted` as you described. What I can see is like this:
`[info] java.lang.RuntimeException: File test/Test1.java does not exist in C:\Users\augustyn\AppData\Local\Temp\sbt_241a615f\basic\target\scala-2.10\src_managed\main`

I removed all the tests except `basic` to shorten the test loop. What I can see:
1. `File test/Test1.java does not exist in C:\Users\augustyn\AppData\Local\Temp\sbt_60f28a65\basic\target\scala-2.10\src_managed\main.` I can see that there is just `test\Test2.java` file in that directory.
2. `File test/Test1.class does not exist in C:\Users\augustyn\AppData\Local\Temp\sbt_60f28a65\basic\target\scala-2.10\classes.`. I can see these files in the `test` subdirectory: `Test2$1.class Test2$test2$1.class Test2$test2$Builder.class Test2$test2.class Test2$test2OrBuilder.class Test2.class`

Where I should look for more detailed logs?

augi commented:
What is strange that I can see in the output:
```
[info] protoc-jar: executing: [C:\Users\augustyn\AppData\Local\Temp\protocjar7593498991892945809\bin\protoc.exe, --java_out=:C:\Users\augustyn\AppData\Local\Temp\sbt_f3a9e8f1\basic\target\scala-2.11\src_managed\main, -IC:\Users\augustyn\AppData\Local\Temp\sbt_f3a9e8f1\basic\src\main\protobuf, -IC:\Users\augustyn\AppData\Local\Temp\sbt_f3a9e8f1\basic\target\protobuf_external, -IC:\Users\augustyn\AppData\Local\Temp\sbt_f3a9e8f1\basic\src\main\protobuf, -IC:\Users\augustyn\AppData\Local\Temp\sbt_f3a9e8f1\basic\target\protobuf_external, C:\Users\augustyn\AppData\Local\Temp\sbt_f3a9e8f1\basic\src\main\protobuf\test2.proto]
```
But I've never seen a command that would process also `test1.proto` file :(

augi commented:
@thesamet I removed my previous comments - they were misleading as I didn't know how `scripted` works. It's still weird for me that it prints the exceptions although the script is marked as `->` (so failure is expected).

Anyway, I executed `sbt ^test ^scripted` (as stated in `.travis.yml`) and it ended with `success`. 

The only change in `sbt-protoc` projects is that I changed the dependency to the my SNAPSHOT:
```sbt
resolvers += "Local Maven Repository" at Path.userHome.asFile.toURI.toURL + ".m2/repository"
libraryDependencies ++= Seq(
  "com.github.os72" % "protoc-jar" % "3.5.1",
  "com.thesamet.scalapb" %% "protoc-bridge" % "0.7.1-SNAPSHOT"
)
```

augi commented:
I've also tried all the examples. I have just an issue with `grpc-java`:
```
C:\Users\augustyn\Documents\sbt-protoc\examples\grpc-java\build.sbt:21: error: object SafeLazy is not a member of package xsbti
grpcExePath := xsbti.SafeLazy {
                     ^
C:\Users\augustyn\Documents\sbt-protoc\examples\grpc-java\build.sbt:25: error: value download is not a member of object sbt.io.IO
    IO.download(grpcExeUrl, exe)
       ^
[error] Type error in expression
```
All others examples succeeded (`sbt test`).

thesamet commented:
Thanks, great progress! For `grpc-java`, can you add a `project/build.properties` for it set to 0.13.17?

augi commented:
ACK, `grpc-java` works if I explicitly set sbt version to use.

thesamet commented:
Merged in bdbf2c9. Also added continuous integration with [appveyor](https://ci.appveyor.com/project/thesamet/protoc-bridge)

thesamet commented:
Released in protoc-bridge 0.7.2, available with sbt-protoc 0.99.15
